### ISSUE TITLE
New message routing concept 

### ISSUE THREAD

**Issue by bmoneke** [pymeasure/leco-protocol#34]
Until now, the concept was to somehow manage a translation list between full IDs of Components and their respective identity (or "address") in the Router socket.

In #33 @bklebel pointed out, that Components can set their identity (before connecting). If that identity is already used at connect time, the connect fails.

Messages to a not existing identity are dropped, unless an error check is set on the router socket.

Now the routing proposal:

Components send a message with the routing information. 
For example CA on Co1sends: `Co2 | CB | | Content` ("|" indicate frame boundaries)
The Co1 receives `CA | CO2 | CB | | Content` (CA is added automatically by its own router socket) and sends to its router socket `CO2 | CB | | Content | | Co1 | CA` (moving the sender and its own identity to the back, separated with an empty frame).
The Dealer socket of `CO2` receives `CB | Content | | Co1 | CA`. And that messages is handed as is to the router socket of `CO2`. 
Finally `CB` receives `| Content | | Co1 | CA`
If `CB` wants to send a response, it takes the appended routing information (separated by empty frame) and uses it as destination:
`Co1 | CA | | Reply `
And the whole route happens in reverse
(Content and reply could be several frames) 

~~~mermaid
flowchart TB
    CA-- "Co2|CB||Content" --> Co1.R

    subgraph Co1
        Co1.R[ROUTER] -- "CA|Co2|CB||Content" ---> Co1.C[Code]
        Co1.C -- "Co2|CB||Content||Co1|CA" --> Co1.R
        Co1.D[DEALER] -. "CA||Reply||Co2|CB" .-> Co1.C
        Co1.C -. "CA||Reply||Co2|CB" .-> Co1.R
    end

    Co1.R -- "CB||Content||Co1|CA" --> Co2.D

    subgraph Co2
        Co2.D[DEALER] -- "CB||Content||Co1|Ca" --> Co2.C[Code]
        Co2.C -- "CB||Content||Co1|Ca" --> Co2.R[ROUTER]
        Co2.R -. "CB||Co1|CA||Reply" .-> Co2.C
        Co2.C -. "Co1|CA||Reply||Co2|CB" .-> Co2.R
    end

    Co2.R -- "||Content||Co1|CA" --> CB

    CB -. "Co1|CA||Reply" .-> Co2.R
    Co2.R -. "CA||Reply||Co2|CB" .-> Co1.D
    Co1.R -. "||Reply|Co2|CB" .-> CA
~~~
Now "|" indicate frame boundaries, which are left out at beginning and end, unless there is an empty frame: `| | Content` contains two frames: an empty one and "Content". Full line represents the initial message, dotted line the response.

@bklebel, what do you think? 

Consequences:
- This requires one Dealer socket per connected Coordinator (symmetric) and not only in one direction, but makes the code simple. 
- We could allow from co1's dealer to co2's router as well, which would half the connection number but increase the code complexity.
- as we use frames for routing (multi hob is possible), the sender is added by the Coordinator and not given by the sender itself
- multi hop would work natively: `CO2 | Co3 | Co4 | CX | ....` Content will arrive at CX via Co4 and via Co3 and CO2 without code change, but CA (the sender) has to know the routing information 
- due to frames routing we could add message and conversation ID into the data payload or as the first frames of the "Content": `CO2 | CB | | Message ID | Conversation ID | Real payload`
- router sockets should raise an error if an identity does not exist
- we do not need anymore to keep address books for routing information (just for curiosity of Components) 

EDITed to give frame boundaries in "|"

**Comment by bmoneke** [comment]
Additional comment:
If the first frame is the Coordinator'own identity, it will drop it:
CA sends Co1. CB.. Content to Co1
Co1 cannot send Co1. CB.. Content.. Co1. CA via its router, as the first frame is its own identity, therefore it sends CB.. Content.. Co1. CA


That would make multi hop easier, as every Coordinator (not just the first one) could add itself to the return routing information.

If every Coordinator does so, we should invert the sequence:

A message to Co1. CO2. Co3. Co4. CB.. Content will change
Coordinator one appends CA. Co1 (because it arrived at the router port) 
Coordinator 2 appends only CO2 (because it arrived at the dealer)... 
Such that .Content.. CA. Co1. CO2. Co3. Co4 arrives at Cb
Now CB has to invert the frames to send its reply as
Co4. Co3. CO2. Co1. CA.. Reply


This will allow multi hop routing without more code (maybe needed for NAT or so)

**Comment by bilderbuchi** [comment]
I'm out of time, so one quick remark for now:
> but CA (the sender) has to know the routing information

I would really like to avoid senders to have to know _routing_ information. That's what we have Coordinators for! The sender has to know the namespace.name of the recipient, and its Coordinator takes care of the rest!

(Imagine if, instead of typing `www.google.com`, you'd have to specify all the individual hops of the packets between you and that server)

Also, maybe use `|` as the frame separator in the discussion? It's a bit more prominent than a dot -- I found it hard to follow with dots and whitespace, maybe `Co1|CO2|Co3|Co4` is clearer?

**Comment by bmoneke** [comment]
> I would really like to avoid senders to have to know routing information. That's what we have Coordinators for! The sender has to know the namespace.name of the recipient, and its Coordinator takes care of the rest!

As long as we have not more than two Coordinators for a single message (as we consider until now), both ideas are the same. Nevertheless, we can put the burden upon the Coordinator.

> Also, maybe use | as the frame separator in the discussion? It's a bit more prominent than a dot -- I found it hard to follow with dots and whitespace, maybe Co1|CO2|Co3|Co4 is clearer?

Good idea.

Here is the decision scheme (bold lines indicate how the message is handed on, thin lines programming decisions):

~~~mermaid
flowchart TB
    C([CA]) == "Namespace|Recipient||Content" ==> R0
    R0[Co1 ROUTER receive] == "CA|Namespace|Recipient||Content" ==> Code[Modify address]
    Code == "Namespace|Recipient||Content||Co1|CA" ==> NS
    NS{Namespace} -- "== Co1"--> Drop
    Drop[Drop namespace] == "Recipient||Content||Co1|CA" ==> R1[send]
    R1 == "||Content" ==> W1([Wire to Recipient])
    NS -- "in connections" --> Keep
    Keep[send as is] == "Namespace|Recipient||Content||Co1|CA" ==> R2[send]
    R2 == "Recipient||Content||Co1|CA" ==> W2([Wire to Namespace])
    NS -- "reachable via CoX" --> Augment
    Augment[Add routing information] == "CoX|Namespace|Recipient||Content||Co1|CA" ==> R3[send]
    R3 == "Namespace|Recipient||Content||Co1|CA" ==> W3([Wire to CoX])
    subgraph Co1 ROUTER socket
        R1 -.- R2
        R2 -.- R3
    end
~~~

Notes:
- All three send methods are the same method (indicated by dots), for better seeing the message flow, I showed these three cases.

**Comment by bklebel** [comment]
> I would really like to avoid senders to have to know routing information. That's what we have Coordinators for! The sender has to know the namespace.name of the recipient, and its Coordinator takes care of the rest!

Yes, absolutely. From a "low-entry-threshold" point of view even more so, the less a user needs to put in, the better. As I imagine it, ideal would be if a user would only have to give any Component a unique name/ID, and specify the "location" of any one control Coordinator (i.e. the IP address, ports should be decided on and more or less fixed, but this is then again another issue, could be put into a configuration if a user needs to avoid some ports and go away from the default ones given), and the system does (most) all the rest. 

> @bklebel, what do you think?

This seems quite nice to me. One thing I am not sure about is to move the sender information to the back, rather than at a more or less fixed-length header at the front, `Co2 | CB | | Content | | Co1 | CA` vs `Co2 | CB  | Co1 | CA | Content |`. Sure, it would make it simpler to natively do multi-hop routing, but if I recall our goals correctly, we are still aiming for a limited number of machines and Devices, also going across NAT of Routers is not really what we are concerned about, I think, so I would expect that, within our current goals, for a message to travel between CA and CB, we would only ever need one hop between two Coordinators Co1 and Co2, although 5 or 10 Coordinators might be present in the setup. 
If somebody really needs to hop across firewalls, they can put special Coordinators on both sides of this firewall, with added special location&identity ID's which are only present for this network trip across the firewall. We can think about whether we want to put something like this directly into the protocol, but I would put it way back to the recently mentioned backlog milestones. 

If we always only ever have one Coordinator and one Component in the sender frames, and one Coordinator and one Component in the recipient frames, then we can go with a fixed header length. 
Similar to the mermaid diagram in @bmoneke's first comment in the issue, every Coordinator has i) one ROUTER socket for a) sending and receiving messages from "their" Components, and b) sending messages to other Coordinators, and ii) one DEALER socket for every other Coordinator it is connected to (DEALER sockets **connect**, ROUTER sockets **bind**). 

Also, as discussed elsewhere, all Coordinators could have a PUB and a SUB socket to connect to a bookkeepers proxy, so that announcements of new Coordinators/Components and regular publications of lists of connected Components can be delivered. Possibly we could do this bookkeeping also in a distributed fashion over the existing DEALER and ROUTER sockets, so that we do not have the potential failure of the bookkeeper (or a forgetful user, who did not start it, I am looking at myself) prevent new Components from connecting to anything - this would just give us one more different type of message over the control-channel, instead of having a separate discovery-channel.

**Comment by bmoneke** [comment]
>   One thing I am not sure about is to move the sender information to the back, rather than at a more or less fixed-length header at the front,

We could do : Recipient information (1 to two frames)| empty frame| Sender information (if applicable)| empty frame|Content. As only the first Coordinator changes the message (adding the Sender information), it is not that important, whether in the middle or in the end.
The only reason I saw in putting the sender information at the back is: You get first the content and only if you want to respond, you do have to concern yourself, whether the message contains sender information.

> Also, as discussed elsewhere, all Coordinators could have a PUB and a SUB socket to connect to a bookkeepers proxy,

I assumed this idea to be present, such that the Coordinators know the other ones. Here I focussed on the way how messages are exchanged among **known** Coordinators. I have a https://github.com/pymeasure/leco-protocol/issues/22#issuecomment-1410156320

> From a "low-entry-threshold" point of view even more so, the less a user needs to put in, the better. As I imagine it, ideal would be if a user would only have to give any Component a unique name/ID, and specify the "location" of any one control Coordinator (i.e. the IP address, ports should be decided on and more or less fixed, but this is then again another issue, could be put into a configuration if a user needs to avoid some ports and go away from the default ones given), and the system does (most) all the rest.

For that reason I want to allow to send messages without the Coordinator address (e.g. "CB|| Content") such that you do not have to give the name of (the only) Coordinator if you want to talk to some Component.

**Comment by bmoneke** [comment]
In my last two messages I already designed a communication scheme in which the Coordinators hand a message to the next Coordinator in a chain, while the Sender only specifies target Coordinator and Recipient.
Yes, we only want to have two Coordinators, but with that scheme, multi-hop is possible without additional logic in the Components.
For multi-hop, only the question remains, how to exactly tell that information to the Coordinators.

**Comment by bmoneke** [comment]
Can we agree on the following?

- The first 0-2 frames indicate the recipient (0 for the Coordinator, 1 for a Node's Component and 2 for any Component, even in the same Node)
- Afterwards is one empty frame
- Somewhere (to be determined in #33 regarding header format) we store the return address (be it the first two data frames, or frames separated by an empty frame, be it the last) as frames
- Above mentioned Coordinator logic and Message routing (except for the place of the return address)

That way, we could tie down the routing mechanism and close this issue.

**Comment by bmoneke** [comment]
In issue #27 we found out, that this routing mechanism is kind of supported, but not with love. It might be dropped at some point (gut feeling).

**Comment by bilderbuchi** [comment]
> * I allowed to give the own Coordinator as a namespace, so you could address CB in the namespace either as "CB" or as "Co1|CB".

:+1:

> * As you see, the message to "CoX" is the same as the original message, therefore CoX may know "Namespace" directly and send it there or he may send it to "CoY", without changing the message. (it will always arrive at a DEALER port, such that the return information will not be changed).

That's neat! What's not clear to me is how "is reachable via CoX" is determined? Is there another routing table or so?

Later:
> The Coordinators need just a dictionary of Namespaces and the next hop necessary to reach that namespace (no difficult logic)
For multi-hop, only the question remains, how to exactly tell that information to the Coordinators, but that can be put off, as it is not a goal.

I agree, can be postponed.

> This seems quite nice to me. One thing I am not sure about is to move the sender information to the back, rather than at a more or less fixed-length header at the front,

I agree. I'D prefera fixed number of header frames at the beginning, this makes it easier to reason about the message structure in various phases. (E.g. if an in-between Coordinator needs to deserialise a message for some reason, determining which frame the payload is could be...involved. Better to know that it's always the 5th or last or whatever.

> Similar to the mermaid diagram in @bmoneke's first comment in the issue, every Coordinator has i) one ROUTER socket for a) sending and receiving messages from "their" Components, and b) sending messages to other Coordinators, and ii) one DEALER socket for every other Coordinator it is connected to (DEALER sockets connect, ROUTER sockets bind).

This sounds like a neat approach (although as you know my zmq experience is limited)

> Also, as discussed elsewhere, all Coordinators could have a PUB and a SUB socket to connect to a bookkeepers proxy, so that announcements of new Coordinators/Components and regular publications of lists of connected Components can be delivered. Possibly we could do this bookkeeping also in a distributed fashion over the existing DEALER and ROUTER sockets, so that we do not have the potential failure of the bookkeeper (or a forgetful user, who did not start it, I am looking at myself) prevent new Components from connecting to anything - this would just give us one more different type of message over the control-channel, instead of having a separate discovery-channel.

I'd prefer having an additional message to having an additional (singleton, to boot!) proxy and connection set. This is simpler, we have to keep track of fewer connections and fewer entities, at the cost of 1 more message type.

**Comment by bilderbuchi** [comment]
> Can we agree on the following?
> 
>     * The first 0-2 frames indicate the recipient (0 for the Coordinator, 1 for a Node's Component and 2 for any Component, even in the same Node)
> 
>     * Afterwards is one empty frame
> 
>     * Somewhere (to be determined in [Control protocol header format #33](https://github.com/pymeasure/leco-protocol/issues/33) regarding header format) we store the return address (be it the first two data frames, or frames separated by an empty frame, be it the last) as frames
> 
>     * Above mentioned Coordinator logic and Message routing (except for the place of the return address)
> 
> 
> That way, we could tie down the routing mechanism and close this issue.

Sounds good to me with what I know right now. (see my comments)

**Comment by bmoneke** [comment]
Status: We reverted to the first idea of using names, not the connection identities, which are not supported that well.

So, this issue should be closed.
